### ISSUE TITLE
Add more progress text styles to ProgressBar node

### ISSUE THREAD

**Issue by wingedadventurer** [godotengine/godot-proposals#2399]
### Describe the project you are working on
Any project that uses ProgressBar node.

### Describe the problem or limitation you are having in your project
ProgressBar node shows only percentage. But two other common styles one might want is to display it as a fraction (ex. health bar with value `230/600`) or decimal (ex. attack modifier with value `0.72`).

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
Add few more options to pick besides percentage. The 2 above are most common I believe, but if anyone has ideas for more, shoot.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
It could be done as an enum:

![proposal](https://user-images.githubusercontent.com/26509794/110009363-2ef0f900-7d1d-11eb-833d-f3f37375e817.png)

### If this enhancement will not be used often, can it be worked around with a few lines of script?
Easiest solution would be to add a child Label node set to Full Rect, and modify it on bar value change.

It would be hard to make this proposal cover all cases, and it's better to use custom label at that point, as mentioned in #322. But I believe that the 2 cases above are quite common, and shouldn't cause any performance issues or add too much complexity.

### Is there a reason why this should be core and not an add-on in the asset library?
It's a change on core node.

**Comment by KoBeWi** [comment]
Kinda late seeing as implementation already exists, but ProgressBar is meant to be used for loading progress mostly. For health bars, TextureProgress with a label would be better. Not sure if adding any more flexibility to ProgressBar is really useful, as users might still not find it enough. I wonder if, instead of adding more modes, a virtual method like `_get_progress_string()` could be better. That way you could set an arbitrary text on top of the bar, which would completely exhaust any customization needs.

Personally I never used ProgressBar though, so not sure.

**Comment by wingedadventurer** [comment]
I agree completely with @KoBeWi. The virtual method sounds like a much better option. It gives full label format freedom.

Fraction and decimal modes ended up looking wonky anyways when I implemented those, because it removes excess decimal zeros so the string size jumps around (ex. from 3.4 to 3.402), and adding extra properties to control those makes it unnecessarily complicated.

The thing now however is that virtual method idea would make it more complicated to implement for a game, as it requires a script attached to that node and overriding the method, instead of a simple enum that's quickly selectable from the editor. From that perspective, a custom label node doesn't seem that more complex to implement.

Personally I use ProgressBar node for simple health bars, and always remove the percentage view as it's ugly.

I can try implementing that virtual method if no other ideas get suggested.

**Comment by Maran23** [comment]
This is something I would like to see implemented as well.

**Comment by pycbouh** [comment]
@Maran23 Please don't bump proposals just to show your support. Use reactions on the first post for that.

**Comment by KoBeWi** [comment]
Well, another idea is a format string. By default it would be `%d%%`, so it displays percentage, but could be changed to have any prefix/suffix or precision and the final string would be `format % current_value`. The problem is that it wouldn't support fraction display, unless more complex placeholder behavior is implemented.

Or the most flexible option - providing 2 options, i.e. format string/enum + virtual method. If you implement virtual method, its result will be used. If not, the format will be used.

**Comment by KoBeWi** [comment]
> Am I right that a virtual method needs to be implemented in a script (while the format does not necessary need one)?

Yes, overriding virtual method requires a script. The format wouldn't need one; it's just a string property.

**Comment by Calinou** [comment]
> The problem is that it wouldn't support fraction display, unless more complex placeholder behavior is implemented.

There could be several named placeholders available for use, similar to `String.format()`'s syntax.

For instance: `{fraction}`, `{value}`, `{max}`

**Comment by KoBeWi** [comment]
String formatting takes array of arguments. So what we could the is "parse" the string when setting the format and remember what placeholders were used, then in the final string format based on these arguments.

E.g. we could have a custom prefix `M` that denotes "max value" so you could e.g. use `"%.2f / % M%.2f"` format, the ProgressBar would find the `M` and remove it then display the value with `"%.2f / %.2f" % [value, max_value]`.

It's complex, but relatively easy to implement. I think this is better suited as an addon though. It's easy to add a label over a ProgressBar and make it display the value, which allows you to do whatever you want.

**Comment by Calinou** [comment]
> It's complex, but relatively easy to implement. I think this is better suited as an addon though. It's easy to add a label over a ProgressBar and make it display the value, which allows you to do whatever you want.

I agree this is also better left to an add-on. Now I'm not sure what would be the best approach in core (especially following [guideline #6](https://docs.godotengine.org/en/stable/community/contributing/best_practices_for_engine_contributors.html#cater-to-common-use-cases-leave-the-door-open-for-the-rare-ones)):

- Virtual method is flexible, but requires attaching a script to your ProgressBar.
- Format string is simple, but limited.

**Comment by KoBeWi** [comment]
Or we could just go with the already implemented solution and if someone wants more they can use an addon.

btw regarding my old comment:
> For health bars, TextureProgress with a label would be better.

As I used ProgressBars a bit more recently, a nice thing about them is that they use styleboxes. So TextureProgress isn't a direct replacement. Still, you can put a Label over progress bar too.
